controllers/k8ssandra/k8ssandracluster_controller.go

```diff
@@ -197,239 +114,61 @@ func (r *K8ssandraClusterReconciler) reconcile(ctx context.Context, kc *api.K8ss
-				logger.Info("Waiting for datacenter to become ready")
-				return ctrl.Result{RequeueAfter: r.ReconcilerConfig.DefaultDelay}, nil
-			}
-
-			logger.Info("The datacenter is ready")
+	kcLogger.Info("All dcs reconciled")
 
-			actualDcs = append(actualDcs, actualDc)
-		} else {
-			if errors.IsNotFound(err) {
-				// cassdc doesn't exist, we'll create it
-				if err = remoteClient.Create(ctx, desiredDc); err != nil {
-					logger.Error(err, "Failed to create datacenter")
-					return ctrl.Result{}, err
-				}
-				return ctrl.Result{RequeueAfter: r.ReconcilerConfig.DefaultDelay}, nil
-			} else {
-				logger.Error(err, "Failed to get datacenter")
-				return ctrl.Result{}, err
-			}
-		}
+	if recResult := r.reconcileStargateAuthSchema(ctx, kc, actualDcs, kcLogger); recResult.Completed() {
+		return recResult.Output()
 	}
 
-	// If we reach this point all CassandraDatacenters are ready. We only set the
-	// CassandraInitialized condition if it is unset, i.e., only once. This allows us to
-	// distinguish whether we are deploying a CassandraDatacenter as part of a new cluster
-	// or as part of an existing cluster.
-	if kc.Status.GetConditionStatus(api.CassandraInitialized) == corev1.ConditionUnknown {
-		now := metav1.Now()
-		(&kc.Status).SetCondition(api.K8ssandraClusterCondition{
-			Type:               api.CassandraInitialized,
-			Status:             corev1.ConditionTrue,
-			LastTransitionTime: &now,
-		})
+	if recResult := r.recocileReaperSchema(ctx, kc, actualDcs, kcLogger); recResult.Completed() {
+		return recResult.Output()
 	}
 
-	kcLogger.Info("All dcs reconciled")
-
-	if kc.HasStargates() {
-		kcLogger.Info("Reconciling Stargate auth schema")
-		dcTemplate := kc.Spec.Cassandra.Datacenters[0]
-		if remoteClient, err := r.ClientCache.GetRemoteClient(dcTemplate.K8sContext); err != nil {
-			return ctrl.Result{}, err
-		} else if err := r.reconcileStargateAuthSchema(ctx, kc, actualDcs[0], remoteClient, kcLogger); err != nil {
-			return ctrl.Result{RequeueAfter: r.ReconcilerConfig.LongDelay}, err
-		}
+	if recResult := r.reconcileStargateAndReaper(ctx, kc, actualDcs, kcLogger); recResult.Completed() {
+		return recResult.Output()
 	}
 
-	if kc.HasReapers() {
-		kcLogger.Info("Reconciling Reaper schema")
-		dcTemplate := kc.Spec.Cassandra.Datacenters[0]
-		if remoteClient, err := r.ClientCache.GetRemoteClient(dcTemplate.K8sContext); err != nil {
-			return ctrl.Result{}, err
-		} else if err := r.reconcileReaperSchema(ctx, kc, actualDcs[0], remoteClient, kcLogger); err != nil {
-			return ctrl.Result{RequeueAfter: r.ReconcilerConfig.LongDelay}, err
-		}
-	}
+	kcLogger.Info("Finished reconciling the k8ssandracluster")
 
-	// Reconcile Stargate and Reaper across all datacenters
+	return result.Done().Output()
+}
+
+func (r *K8ssandraClusterReconciler) reconcileStargateAndReaper(ctx context.Context, kc *api.K8ssandraCluster, dcs []*cassdcapi.CassandraDatacenter, logger logr.Logger) result.ReconcileResult {
```

Comment:
Maybe rename the method to something like `afterDatacentersFullyReconciled` since we are likely going to add more stuff here (Medusa, for instance).

Reply:
I'm good with renaming it, but Medusa stuff won't go there. The main parts for Medusa are deployed in the Cassandra pod.

Reply:
The more I think about it I am not sure about the change. If I make that name change then `reconcileStargateAuthSchema` and `reconcileReaperSchema` should also be called in that method. Maybe that's fine. I suppose it would make it clear about where to put new method/function calls for later changes. @adutra wdyt?

Reply:
Indeed, schema reconciliation would have to happen inside this method. Since Medusa won't be involved here, I'd say let's leave the name as it is for now.